### PR TITLE
feat: Support publishing source bundles to chain and installing hash bundles to Zoe

### PR DESCRIPTION
**Pull request by kriskowal**
This stack of changes-in-progress will address:

- #4974
- #4396
- #4564
- #4565

For manual verification, clone `dapp-fungible-faucet`, check out `publish-bundle`, and run these commands in separate sessions:

```
agoric start local-solo
agoric start local-chain
DEBUG=agoric:deploy agoric deploy ./contract/deploy.js
```

### CODE REVIEW

File: packages/agoric-cli/src/deploy.js [hunk 1]
```diff
@@ -231,11 +232,12 @@ export { bootPlugin } from ${JSON.stringify(absPath)};
...
           // Use Node.js ESM support if package.json of template says "type":
           // "module".
-          const read = async url => fs.readFile(new URL(url).pathname);
+          const read = async location =>
+            fs.readFile(url.fileURLToPath(location));
           const { packageDescriptorText } =
             await readContainingPackageDescriptor(
               read,
-              `file://${moduleFile}`,
+              url.pathToFileURL(moduleFile),
```
Comment (dckc) [inline packages/agoric-cli/src/deploy.js:369]: something tells me "portable" also means powerful / ambient

File: packages/publish-bundle/LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (dckc) [inline packages/publish-bundle/LICENSE]: How did you scaffold this new project? Would you please leave some clues in the commit message?

File: packages/agoric-cli/src/deploy.js (hunk 1 above)
Comment (kriskowal) [reply packages/agoric-cli/src/deploy.js:369]: It certainly at least means “sensitive to platform”. The big difference is the treatment of drive letters as absolute paths instead of concatenating them like `c:/users/kris/endo/d:/users/kris/` etc.

File: packages/publish-bundle/LICENSE (hunk 2 above)
Comment (kriskowal) [reply packages/publish-bundle/LICENSE]: I’ll leave a note. It’s only semi automatic.

File: packages/agoric-cli/src/deploy.js [hunk 3]
```diff
@@ -1,17 +1,23 @@
-/* global process setTimeout setInterval clearInterval */
+/* eslint-env node */
```
Comment (kriskowal) [inline packages/agoric-cli/src/deploy.js]: Spread the word.

File: packages/agoric-cli/src/deploy.js [hunk 4]
```diff
@@ -89,9 +103,60 @@ export default async function deployMain(progname, rawArgs, powers, opts) {
...
+    const { connectionNumber } = await inquirer.prompt({
+      name: 'connectionNumber',
+      message: 'Please choose a connection to publish to',
+      choices: lookup.keys(),
+    });
+    const connection = lookup.get(connectionNumber);
+    console.log({ connection });
+    return connection;
+  };
+
```
Comment (kriskowal) [inline packages/agoric-cli/src/deploy.js]: - This is interactive
- Perhaps it should not prompt if there is 1 option
- Perhaps it should err if there are 0 options
- Perhaps interactive mode should be overridden by a command line option, if present, but describing connections requires one of various combinations of flags

Please make your preference known :-) @michaelfig @warner

File: packages/agoric-cli/src/deploy.js [hunk 5]
```diff
@@ -89,9 +103,60 @@ export default async function deployMain(progname, rawArgs, powers, opts) {
...
+    cosmosHelper,
+    pathResolve: path.resolve,
+    writeFile: fs.writeFile,
+  });
+  const publishBundle = makeBundlePublisher({
+    getAccessToken,
+    jsonHttpCall,
+    publishBundleCosmos,
+    getDefaultConnection,
+  });
```
Comment (kriskowal) [inline packages/agoric-cli/src/deploy.js]: I’m a little dissatisfied that the HTTP guts are inside `@agoric/publish-bundle` but the Cosmos SDK guts are in `@agoric/agoric`. Putting it all here makes it harder to test in `@agoric/solo`, which doesn’t work in `solo/test/test-home.js` yet anyway. Putting all the guts in `@agoric/publish-bundle` would require some duplication or refactoring of the utilities above that are needed to make the bundle publisher.

File: packages/agoric-cli/src/deploy.js [hunk 6]
```diff
@@ -242,11 +307,12 @@ export { bootPlugin } from ${JSON.stringify(absPath)};
 
           // Use Node.js ESM support if package.json of template says "type":
           // "module".
-          const read = async url => fs.readFile(new URL(url).pathname);
+          const read = async location =>
+            fs.readFile(url.fileURLToPath(location));
```
Comment (kriskowal) [inline packages/agoric-cli/src/deploy.js:365]: This is a drive-by portability fix for Windows.

File: packages/agoric-cli/src/publish.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      const tempFilePath = pathResolve(tempDirPath, 'bundle.json');
+      await writeFile(tempFilePath, `${JSON.stringify(bundle)}\n`);
+      const args = [
+        'tx',
+        'swingset',
+        'install-bundle',
+        '--gas',
+        'auto',
+        '--gas-adjustment',
+        '1.2',
```
Comment (kriskowal) [inline packages/agoric-cli/src/publish.js:185]: I definitely want to put off parameterizing these.

File: packages/agoric-cli/src/start.js [hunk 8]
```diff
@@ -73,6 +73,7 @@ export default async function startMain(progname, rawArgs, powers, opts) {
 
   const sdkPrefixes = {};
   if (!opts.sdk) {
+    // TODO importMetaResolve should be more reliable.
```
Comment (kriskowal) [inline packages/agoric-cli/src/start.js]: Note to self: remove this comment and track elsewhere.

File: packages/cosmic-swingset/src/block-manager.js [hunk 9]
```diff
@@ -79,6 +80,17 @@ export default function makeBlockManager({
         break;
       }
 
+      case ActionType.INSTALL_BUNDLE: {
+        p = (async () => {
+          const bundle = JSON.parse(action.bundle);
+          harden(bundle);
+          return validateAndInstallBundle(bundle);
+        })().catch(error => {
+          console.error(error);
```
Comment (kriskowal) [inline packages/cosmic-swingset/src/block-manager.js:89]: I imagine this error message would be opaque without some more context. Not sure what would be useful to the end user.

File: packages/publish-bundle/node-powers.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+/* eslint-env node */
+
+const textEncoder = new TextEncoder();
+const textDecoder = new TextDecoder();
+
+/**
+ * @param {Object} powers
+ * @param {typeof import('http')} powers.http
+ */
+export const makeJsonHttpClient = ({ http }) => {
```
Comment (kriskowal) [inline packages/publish-bundle/node-powers.js]: I wrote this so it could be faked in tests and to avoid coupling to the powerful `http`.

File: packages/solo/src/chain-cosmos-sdk.js [hunk 11]
```diff
@@ -21,10 +22,9 @@ const console = anylogger('chain-cosmos-sdk');
 // the `agd` tool in our repo is built by 'cd golang/cosmos &&
 // make'. It lives in the build tree along with `bin/ag-solo`, in case there are
 // multiple checkouts of `agoric-sdk`.
-export const HELPER = new URL(
-  '../../../golang/cosmos/build/agd',
-  import.meta.url,
-).pathname;
+export const HELPER = url.fileURLToPath(
+  new URL('../../../golang/cosmos/build/agd', import.meta.url),
+);
```
Comment (kriskowal) [inline packages/solo/src/chain-cosmos-sdk.js:27]: Again with a drive-by fix for Windows portability.

File: packages/solo/src/web.js [hunk 12]
```diff
@@ -10,6 +10,10 @@ import morgan from 'morgan';
 
 import { getAccessToken } from '@agoric/access-token';
 
+const textDecoder = new TextDecoder();
+
+const maximumBundleSize = 1024 * 1024 * 128; // 128MB
```
Comment (kriskowal) [inline packages/solo/src/web.js:13]: Arbitrary. Open to suggestions for a less arbitrary mechanism or a less arbitrary magic number.

File: packages/solo/src/web.js [hunk 13]
```diff
@@ -194,6 +200,66 @@ export async function makeHTTPListener(
...
+      return;
+    }
+
+    res.json({
+      ok: true,
+      connections,
+    });
+  });
+
+  app.post('/publish-bundle', async (req, res) => {
```
Comment (kriskowal) [inline packages/solo/src/web.js:213]: This is not tested. It would be tested by `solo/test/test-home.js` if that worked. There’s probably something wrong here with request or response size limits, despite the limit set above. However, this is just for the `http` connection type and isn’t employed when submitting to a Cosmos chain, fake or otherwise.

File: packages/solo/test/test-home.js [hunk 14]
```diff
@@ -1,16 +1,19 @@
 /* global process */
 
-import '@endo/init/pre-bundle-source.js';
-
 // `test.after.always` does not yet seem compatible with ses-ava
 // See https://github.com/endojs/endo/issues/647
 // TODO restore
-// import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava';
```
Comment (kriskowal) [inline packages/solo/test/test-home.js:8]: Drive-by fix for the `.js` extension, left over before the RESM-to-NESM migration.

File: packages/solo/test/test-home.js [hunk 15]
```diff
@@ -1,16 +1,19 @@
 /* global process */
 
-import '@endo/init/pre-bundle-source.js';
```
Comment (kriskowal) [inline packages/solo/test/test-home.js:3]: We don’t need this anymore, since Endo isn’t allergic to its own Babel anymore.

File: packages/solo/test/test-home.js [hunk 16]
```diff
@@ -1,16 +1,19 @@
...
-
 // `test.after.always` does not yet seem compatible with ses-ava
 // See https://github.com/endojs/endo/issues/647
 // TODO restore
-// import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava';
+// import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
 import '@agoric/swingset-vat/tools/prepare-test-env.js';
 // eslint-disable-next-line import/no-extraneous-dependencies
 import test from 'ava';
+// TODO
```
Comment (kriskowal) [inline packages/solo/test/test-home.js]: Uncomment all the TODOs in this file to reproduce the publishBundle HTTP failure. Otherwise, this should be sufficient to test the HTTP connection type in `@agoric/publish-bundle`, but not sufficient to test for regressions in publishing to a Cosmos chain. I’m not sure where to hook up such a test.

File: packages/zoe/src/contractFacet/internal-types.js [hunk 17]
```diff
@@ -20,7 +20,7 @@
 
 /**
  * @typedef ZCFZygote
- * @property {(bundleOrBundleCap: SourceBundle | BundleCap) => void} evaluateContract
+ * @property {(bundleOrBundleCap: Record<string, string>) => void} evaluateContract
```
Comment (kriskowal) [inline packages/zoe/src/contractFacet/internal-types.js]: `SourceBundle` wasn’t defined any more specifically than this. It probably should be `unknown` since bundles get trafficked as JSON and probably should be verified anywhere they’re bought or sold. I could easily be swayed to go the other way, making these type unions hyper explicit, but that seemed out of scope for this change.

File: packages/zoe/src/zoeService/zoe.js [hunk 18]
```diff
@@ -62,7 +62,8 @@ const makeZoeKit = (
     shutdownZoeVat,
   );
 
-  const getBundleCapFromID = bundleID => E(vatAdminSvc).getBundleCap(bundleID);
+  const getBundleCapFromID = bundleID =>
+    E(vatAdminSvc).waitForBundleCap(bundleID);
```
Comment (kriskowal) [inline packages/zoe/src/zoeService/zoe.js:66]: @warner This was the most surgical change I could make to ensure that `E(zoe).install(hashBundle)` would wait for the chain to receive the bundle separately. This might be giving us a hint that `getBundleCap` should behave like `waitForBundleCap`, rather than having separate methods, or it might be a hint that `getBundleCapFromID` defined here should be renamed `waitForBundleCapWithID` or some such.

File: packages/zoe/tools/fakeVatAdmin.js [hunk 19]
```diff
@@ -56,6 +56,12 @@ function makeFakeVatAdmin(testContextSetter = undefined, makeRemote = x => x) {
       }
       return Promise.resolve(idToBundleCap.get(bundleID));
     },
+    waitForBundleCap: bundleID => {
+      if (!idToBundleCap.has(bundleID)) {
+        idToBundleCap.init(bundleID, fakeBundleCap());
+      }
+      return Promise.resolve(idToBundleCap.get(bundleID));
+    },
```
Comment (kriskowal) [inline packages/zoe/tools/fakeVatAdmin.js:64]: @warner For the fake, I duplicated `getBundleCap`. This might be another hint that I should just change the behavior of the real `getBundleCap`.

File: packages/agoric-cli/src/publish.js [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+        chainID,
+        '--yes',
+        `@${tempFilePath}`,
+      ];
+      await pspawn(cosmosHelper, args);
+    } finally {
+      removeCallback();
+    }
+  };
+
```
Comment (kriskowal) [inline packages/agoric-cli/src/publish.js:202]: @michaelfig This would be a good place to wait for an event from the chain telling us whether the INSTALL_BUNDLE action succeeded or failed, probably by laying a trap somewhere keyed on the hash of the bundle we attempted to install. I think I’ll need you looking over my shoulder for that project.

Otherwise, this hangs indefinitely if publishing fails.

File: packages/cosmic-swingset/src/block-manager.js [hunk 21]
```diff
@@ -79,6 +80,17 @@ export default function makeBlockManager({
         break;
       }
 
+      case ActionType.INSTALL_BUNDLE: {
+        p = (async () => {
+          const bundle = JSON.parse(action.bundle);
+          harden(bundle);
+          return validateAndInstallBundle(bundle);
```
Comment (kriskowal) [inline packages/cosmic-swingset/src/block-manager.js:87]: After this would be a good time to emit an “event” as to whether the installation worked or not. My understanding from @michaelfig is that this could arrive in the Agoric deploy script via its WebSocket connection, so `publishBundle` would need a mechanism to connect a listener for a particular bundle hash’s installBundle resolution on chain.

File: packages/agoric-cli/src/deploy.js (hunk 3 above)
Comment (dckc) [reply packages/agoric-cli/src/deploy.js]: hm. I kinda like the explicit list.

**Comment by kriskowal** [comment]
Commits are individually reviewable.

File: packages/agoric-cli/src/deploy.js (hunk 3 above)
Comment (kriskowal) [reply packages/agoric-cli/src/deploy.js]: Oh, indeed. I must have been cross-eyed when I wrote this because I thought that the environment might provide a better hint than `any` types for all these globals to TypeScript, but this is an _eslint_ pragma so it probably does not.

**Review by dckc — COMMENTED** [review]
I browsed it; I don't see any problems.

I hope to try it out soonish.

**Review by turadg — APPROVED** [review]
Read commit by commit. I don't know much about bundling mechanics but the code LGTM.

One request: the new package.json has older deps than the rest of agoric-sdk; update those so yarn.lock doesn't take on new versions unnecessarily.

**Comment by dckc** [comment]
What would it cost to measure code coverage?

File: packages/agoric-cli/src/deploy.js (hunk 4 above)
Comment (michaelfig) [reply packages/agoric-cli/src/deploy.js]: Thanks for asking!  I'd like a flag describing the desired installation target:

* `--target=local` for the local solo's own kernel (targeting spawner)
* `--target=cosmos` for the cosmos-based chain
* `--target=sim` for the sim-chain
* and an alias `--target=agoric` for the union of `cosmos` and `sim` connections

I think the default should be `--target=agoric`.

Then, choose the connection automatically if `--target` is specified and a single connection matches.    Otherwise, if stdin is a TTY, give a prompt for all connections, moving any ones matching `--target` to be first in the list.  The final fallback if stdin is not a TTY is to print an error message and exit.

File: packages/zoe/src/contractFacet/internal-types.js (hunk 17 above)
Comment (turadg) [reply packages/zoe/src/contractFacet/internal-types.js]: >  I could easily be swayed to go the other way, making these type unions hyper explicit, but that seemed out of scope for this change.

I'd like to help sway you. If it's out of scope, please leave the existing union as this change is a loss of information. (I take it there are still types, declared or not, `SourceBundle` and `BundleCap` and this `bundleOrBundleCap` is still their union.) While `SourceBundle` is not well typed, at least it's named and there's a ticket to improve it:
https://github.com/Agoric/agoric-sdk/blob/454afd34274017175db5ac0513d6cee5b7375a3a/packages/zoe/src/zoeService/utils.d.ts#L4-L6

`BundleCap` needs typing too of course
https://github.com/Agoric/agoric-sdk/blob/454afd34274017175db5ac0513d6cee5b7375a3a/packages/SwingSet/src/types-external.js#L329

File: golang/cosmos/x/swingset/keeper/msg_server.go [hunk 22]
```diff
@@ -155,3 +155,29 @@ func (keeper msgServer) Provision(goCtx context.Context, msg *types.MsgProvision
 
 	return &types.MsgProvisionResponse{}, nil
 }
+
+type installBundleAction struct {
+	*types.MsgInstallBundle
+	Type        string `json:"type"` // INSTALL_BUND
```
Comment (michaelfig) [inline golang/cosmos/x/swingset/keeper/msg_server.go]:
```suggestion
	Type        string `json:"type"` // INSTALL_BUNDLE
```

File: packages/agoric-cli/src/deploy.js (hunk 3 above)
Comment (michaelfig) [inline packages/agoric-cli/src/deploy.js]:
```suggestion
/* global process setTimeout setInterval clearInterval */
```

File: packages/agoric-cli/src/deploy.js [hunk 23]
```diff
@@ -89,9 +103,60 @@ export default async function deployMain(progname, rawArgs, powers, opts) {
     1000,
   );
 
-  const retryWebsocket = async () => {
-    const accessToken = await getAccessToken(opts.hostport);
+  const accessToken = await getAccessToken(opts.hostport);
```
Comment (michaelfig) [inline packages/agoric-cli/src/deploy.js]: For robustness, the `accessToken` should be refetched on every retry of the websocket.

File: packages/agoric-cli/src/publish.js (hunk 20 above)
Comment (michaelfig) [reply packages/agoric-cli/src/publish.js:202]: The correct way to do that is via the WebSocket, as done for `swingset deliver` transactions in `chain-cosmos-sdk.js`.

File: packages/cosmic-swingset/src/block-manager.js (hunk 21 above)
Comment (michaelfig) [reply packages/cosmic-swingset/src/block-manager.js:87]: Actually, after succeeding in JS, the event should be emitted from Golang, and tracked in deploy-bundle via a Tendermint RPC WebSocket subscription (not the ag-solo WebSocket).

File: packages/solo/src/web.js [hunk 24]
```diff
@@ -1,4 +1,4 @@
-/* global setTimeout clearTimeout setInterval clearInterval */
+/* eslint-env node */
```
Comment (michaelfig) [inline packages/solo/src/web.js]:
```suggestion
/* global setTimeout clearTimeout setInterval clearInterval */
```

File: yarn.lock [hunk 25]
```diff
@@ -14727,6 +14753,11 @@ prettier-linter-helpers@^1.0.0:
   dependencies:
     fast-diff "^1.1.2"
 
+prettier@^1.19.1:
```
Comment (michaelfig) [inline yarn.lock]: Something in your changes is loading the older prettier.  Please correct.

**Review by michaelfig — CHANGES_REQUESTED** [review]
A few comments, and as @turadg suggested, please don't remove the `SourceBundle` type.

File: packages/solo/package.json [hunk 26]
```diff
@@ -33,7 +33,6 @@
     "@agoric/telemetry": "^0.2.0",
     "@agoric/vats": "^0.9.0",
     "@agoric/wallet": "^0.13.0",
-    "@agoric/xsnap": "^0.12.0",
```
Comment (michaelfig) [inline packages/solo/package.json:36]: Why?

**Review by michaelfig — COMMENTED** [review]
Looks great!  Please fix the lint.

File: packages/solo/package.json (hunk 26 above)
Comment (kriskowal) [reply packages/solo/package.json:36]: This is a drive-by. There is no motivation except that xsnap is not uttered anywhere in the solo package, so it appears to be extraneous.

**Review by michaelfig — APPROVED** [review]
My phone is not cooperative.  Lgtm.

File: packages/solo/src/web.js [hunk 27]
```diff
@@ -194,6 +198,33 @@ export async function makeHTTPListener(
...
+  });
+
+  app.post('/publish-bundle', async (req, res) => {
+    if (!(await validateAccessToken(req))) {
+      res.json({ ok: false, rej: 'Unauthorized' });
+      return;
+    }
+    try {
+      const bundle = harden(req.body);
+      await validateAndInstallBundle(bundle);
```
Comment (warner) [inline packages/solo/src/web.js:220]: I'm not sure how it should work / already works in the solo machine, but after doing `validateAndInstallBundle`, the host app is supposed to commit the DB changes. Did you run into anything that might do a swingstore commit?

File: packages/solo/package.json (hunk 26 above)
Comment (warner) [reply packages/solo/package.json:36]: yeah I think swingset has that dependency for you

File: golang/cosmos/x/swingset/types/msgs.go [hunk 28]
```diff
@@ -183,3 +184,32 @@ func (msg MsgProvision) GetSignBytes() []byte {
...
+
+// Type should return the action
+func (msg MsgInstallBundle) Type() string { return "installBundle" }
+
+// ValidateBasic runs stateless checks on the message
+func (msg MsgInstallBundle) ValidateBasic() error {
+	if msg.Submitter.Empty() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "Submitter address cannot be empty")
+	}
+	if len(msg.Bundle) == 0 {
```
Comment (warner) [inline golang/cosmos/x/swingset/types/msgs.go:206]: I'm guessing it's not sufficiently convenient+useful to check e.g. JSON-shape, expected keys, base64-ness, or deeper?

I suppose the only benefit of additional checks would be that I might pay for an installation of something bogus, and get it into a block, only to get rejected after that block was executed by validators.

Eventually I think we want to do more extensive checks here to prevent abuse of blocks as a storage mechanism, catch the misshapen junk before it can get added to a block or flooded to all nodes.

File: packages/cosmic-swingset/src/block-manager.js (hunk 9 above)
Comment (warner) [reply packages/cosmic-swingset/src/block-manager.js:89]: yeah, it might be worthwhile logging the bundle ID that failed to install, so a validator operator could correlated it with the installation txn

and/or the "event" we emit with success/failure should include a copy of this `error` message

File: packages/cosmic-swingset/src/block-manager.js [hunk 29]
```diff
@@ -79,6 +80,17 @@ export default function makeBlockManager({
         break;
       }
 
+      case ActionType.INSTALL_BUNDLE: {
```
Comment (warner) [inline packages/cosmic-swingset/src/block-manager.js:83]: BTW my note about "commit swingstore state" on the `solo` side is not a concern here, the block-manager is already in a context where the state will be committed once block processing is complete

File: packages/zoe/src/zoeService/zoe.js (hunk 18 above)
Comment (warner) [reply packages/zoe/src/zoeService/zoe.js:66]: I think this is fine. I vaguely remember thinking that Zoe's `E(zoe).install()` should wait for the bundle to be installed out-of-band, so that you can only get an installation handle for installations that are actually functional, and aren't waiting for a bundle install that might never happen. And since `E(zoe).install()` is interacting with a wrapper function (which sometimes comes from the fake vatadmin service), that wrapper function should do the waiting, which means that when it calls the real vatadmin service, it should be using `waitForBundleCap` instead of `getBundleCap`.

I've assumed that other (non-zoe, non-agoric) uses/users of swingset and `vatAdminService` may have legitimate reasons to query for a bundleID and want a negative answer (instead of an unbounded pause), which means keeping `getBundleCap` as the form that doesn't wait, and Zoe can use `waitForBundleCap` (which does).

And yeah, at some point renaming this zoe thing `waitForBundleCapFromID` might make sense.

Incidentally, while investigating this function, I found it is named `getBundleCapFromID` here and in `installationStorage.js`, but `getBundleCapForID` in `zoe.js`, which is a bit confusing for code spelunkers.

**Review by warner — APPROVED** [review]
Looks good! Only minor comments.

File: packages/cosmic-swingset/src/block-manager.js (hunk 9 above)
Comment (kriskowal) [reply packages/cosmic-swingset/src/block-manager.js:89]: I think this isn’t the right place to annotate the error with the bundle ID, though it already is annotated through many paths, and it wouldn’t hurt to make sure it’s on every pertinent one. I looked into this and ran into a chicken-and-egg with JSON.parse being a necessary step to see the bundle ID here.

Emitting an event here would be nice too, but I’m going to call it a future enhancement.

File: packages/zoe/src/contractFacet/internal-types.js (hunk 17 above)
Comment (kriskowal) [reply packages/zoe/src/contractFacet/internal-types.js]: I was swayed and the types have been restored.

File: golang/cosmos/x/swingset/types/msgs.go (hunk 28 above)
Comment (kriskowal) [reply golang/cosmos/x/swingset/types/msgs.go:206]: I believe keeping the bundle format opaque at this layer keeps with the spirit of Michael’s intention for the type to evolve freely.
